relay-server/src/actors/global_config.rs

```diff
@@ -253,25 +251,55 @@ impl Service for GlobalConfigService {
-                // for now.
-                self.update_global_config();
-            } else {
-                // NOTE(iker): not making a request results in the sleep handler
-                // not being reset, so no new requests are made.
-                relay_log::info!("fetching global configs disabled: no credentials configured");
-            }
+            match (
+                config.relay_mode(),
+                config.has_credentials(),
+                config.global_config(),
+            ) {
+                (RelayMode::Managed, true, global_config) => {
+                    if global_config.is_some() {
+                        relay_log::info!("ignoring static global config in managed mode");
+                    }
+                    relay_log::info!("serving global configs fetched from upstream");
+                    // This request will trigger the request intervals when internal_rx receives the
+                    // result from upstream.
+                    Self::request_global_config(self.upstream.clone(), self.internal_tx.clone());
+                }
+                (RelayMode::Managed, false, _) => {
+                    // NOTE(iker): not making a request results in the sleep handler
+                    // not being reset, so no new requests are made.
+                    relay_log::info!("serving default global configs due to lacking credentials");
+                }
+                (
+                    RelayMode::Proxy | RelayMode::Static | RelayMode::Capture,
+                    _,
+                    Some(global_config),
+                ) => {
+                    relay_log::info!("serving static global config loaded from file");
+                    self.global_config_watch.send(global_config.clone()).ok();
+                }
+                (RelayMode::Proxy | RelayMode::Static | RelayMode::Capture, _, None) => {
+                    relay_log::info!(
+                        "serving default global configs due to lacking static global config file"
+                    );
+                }
+            };
 
             loop {
                 tokio::select! {
                     biased;
-
-                    () = &mut self.fetch_handle => self.update_global_config(),
-                    Some(result) = self.internal_rx.recv() => self.handle_result(result),
+                    () = &mut self.fetch_handle => {
+                        Self::request_global_config(self.upstream.clone(), self.internal_tx.clone());
+                        // Disable new requests interval until we receive the result from upstream.
+                        self.fetch_handle.reset();
+                    }
+                    Some(result) = self.internal_rx.recv() => {
+                        self.handle_upstream_result(result);
+                        // Start timer for making new global config request.
+                        self.schedule_fetch();
```

Comment:
I moved the timer reset and scheduling away from the functions to make it more explicit, I think it seems a lot more clear here, lmk if any disagreements